go/pkg/cache/clustering/cluster_cache.go

```diff
@@ -0,0 +1,237 @@
+	// Local cache instance to wrap
+	LocalCache cache.Cache[K, V]
+
+	// Topic for broadcasting invalidations
+	Topic *eventstream.Topic[*cachev1.CacheInvalidationEvent]
+
+	// Dispatcher routes invalidation events to this cache
+	// Required for receiving invalidations from other nodes
+	Dispatcher *InvalidationDispatcher
+
+	// Logger for debugging and error reporting
+	Logger logging.Logger
+
+	// Optional node ID (defaults to hostname)
+	NodeID string
+
+	// KeyToString converts a cache key to string for invalidation events
+	// If not provided, fmt.Sprintf("%v", key) will be used
+	KeyToString func(K) string
+
+	// StringToKey converts a string from invalidation events back to key type
+	// If not provided, will attempt to cast string to K
+	StringToKey func(string) (K, error)
+}
+
+// New creates a new ClusterCache that automatically handles
+// distributed cache invalidation across cluster nodes.
+func New[K comparable, V any](config Config[K, V]) (*ClusterCache[K, V], error) {
+	// Validate required config
+	err := assert.All(
+		assert.NotNilAndNotZero(config.Topic, "Topic is required for ClusterCache"),
+		assert.NotNilAndNotZero(config.Dispatcher, "Dispatcher is required for ClusterCache"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set default key converters if not provided
+	keyToString := config.KeyToString
+	if keyToString == nil {
+		keyToString = func(k K) string {
+			if stringer, ok := any(k).(interface{ String() string }); ok {
+				return stringer.String()
+			}
+
+			return fmt.Sprintf("%v", k)
+		}
+	}
+
+	stringToKey := config.StringToKey
+	if stringToKey == nil {
+		stringToKey = func(s string) (K, error) {
+			// Try direct cast for string keys
+			if key, ok := any(s).(K); ok {
+				return key, nil
+			}
+
+			var zero K
+			return zero, fmt.Errorf("cannot convert string %q to key type %T", s, zero)
+		}
+	}
+
+	c := &ClusterCache[K, V]{
+		localCache:  config.LocalCache,
+		topic:       config.Topic,
+		cacheName:   config.LocalCache.Name(),
+		nodeID:      config.NodeID,
+		logger:      config.Logger,
+		keyToString: keyToString,
+		stringToKey: stringToKey,
+		onInvalidation: func(ctx context.Context, key K) {
+			config.LocalCache.Remove(ctx, key)
+		},
+	}
+
+	// Create a reusable producer from the topic
+	c.producer = config.Topic.NewProducer()
+
+	// Register with dispatcher to receive invalidation events
+	config.Dispatcher.Register(c)
+
+	return c, nil
+}
+
+// Get retrieves a value from the local cache
+func (c *ClusterCache[K, V]) Get(ctx context.Context, key K) (value V, hit cache.CacheHit) {
+	return c.localCache.Get(ctx, key)
+}
+
+// Set stores a value in the local cache and broadcasts an invalidation event
+// to other nodes in the cluster
+func (c *ClusterCache[K, V]) Set(ctx context.Context, key K, value V) {
+	// Update local cache first
+	c.localCache.Set(ctx, key, value)
+
+	// Broadcast invalidation to other nodes
+	c.broadcastInvalidation(ctx, key)
```

Comment:
should we actually broadcast here?
setting a value doesn't mean its stale or changed

Reply:
true, I removed it. On Null we are still broadcasting since that would most likely mean something should 404. e.g deleted key / api etc.